# 🍕 음식 배달 주문 데이터 분석

---

## 프로젝트 개요

### 배경
국내 배달 시장은 최근 몇 년간 폭발적으로 성장하며, 배달 플랫폼의 데이터 활용 역량이 핵심 경쟁력이 되었습니다.  
본 프로젝트에서는 가상의 배달 플랫폼 데이터를 활용하여 **매출, 고객, 배달 성과, 날씨 영향** 등을 종합적으로 분석합니다.

### 목표
1. 데이터 전처리를 통해 **분석 가능한 품질**의 데이터를 확보한다.
2. 다양한 관점(매출, 시간, 메뉴, 배달, 고객, 날씨)에서 **인사이트**를 도출한다.
3. 도출된 인사이트를 바탕으로 **비즈니스 제안**을 제시한다.

### 데이터 설명

| 파일 | 설명 | 주요 컬럼 | 행 수 |
|------|------|-----------|-------|
| `stores.csv` | 가게 정보 | 가게ID, 가게명, 업종, 지역, 평균배달시간, 가게평점 | 100 |
| `customers.csv` | 고객 정보 | 고객ID, 성별, 연령, 지역, 가입일, 등급 | 500 |
| `orders.csv` | 주문 정보 | 주문ID, 주문일시, 고객ID, 가게ID, 메뉴, 주문금액, 수량, 배달소요시간, 결제방법, 쿠폰사용여부, 평점, 주문상태 | 5,000 |
| `weather.csv` | 날씨 정보 | 날짜, 기온, 강수량, 날씨 | 366 |

### 분석 프로세스
```
데이터 로딩 → 품질 확인 → 전처리 → 통합 → 탐색적 분석(EDA) → 인사이트 도출 → 제안
```


---
## 1. 환경 설정 & 데이터 로딩


### TODO: 환경 설정

아래 순서대로 환경을 설정하세요.

1. **라이브러리 임포트**: 아래 라이브러리를 불러오세요
   - `pandas as pd`, `numpy as np`
   - `matplotlib.pyplot as plt`, `matplotlib.ticker as ticker`
   - `seaborn as sns`
   - `from matplotlib import font_manager, rc`
   - `warnings`, `os`

2. **경고 무시**: `warnings.filterwarnings('ignore')`

3. **한글 폰트 설정** (OS 자동 분기):
   사용할 메서드: `system()`

4. **마이너스 기호 깨짐 방지**: `plt.rcParams['axes.unicode_minus'] = False`

5. **그래프 스타일 설정**:
   - `plt.rcParams['figure.figsize'] = (12, 6)`
   - `plt.rcParams['figure.dpi'] = 100`
   - `sns.set_style('whitegrid')`

6. **컬러 팔레트 정의**:
   사용할 메서드: `set_palette()`

7. **데이터 경로**: `DATA_DIR = "../data/"`

8. **확인 출력**: 운영체제, pandas/numpy 버전을 출력하세요

**기대 출력**: "환경 설정 완료!", 운영체제명, 라이브러리 버전


In [1]:
# 아래에 코드를 작성하세요
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import koreanize_matplotlib

### TODO: 데이터 로딩

아래 4개 파일을 `pd.read_csv()`로 읽어오세요. 경로는 `os.path.join(DATA_DIR, 파일명)`을 사용합니다.

| 변수명 | 파일 |
|--------|------|
| `stores` | `stores.csv` |
| `customers` | `customers.csv` |
| `orders` | `orders.csv` |
| `weather` | `weather.csv` |

읽은 후 각 데이터프레임의 행 수와 열 수를 출력하세요.



**기대 출력**: stores 100행, customers 500행, orders 5000행, weather 366행


In [2]:
# 아래에 코드를 작성하세요
df_stores = pd.read_csv('./data/stores.csv')
df_customers = pd.read_csv('./data/customers.csv')
df_orders = pd.read_csv('./data/orders.csv')
df_weather = pd.read_csv('./data/weather.csv')

In [3]:
df_stores['업종'].unique()

array(['패스트푸드', '샐러드', '일식', '야식', '분식', '한식', '치킨', '카페/디저트', '중식', '피자'],
      dtype=object)

In [4]:
df_stores[ df_stores['업종'] == '야식' ]

,가게ID,가게명,업종,지역,평균배달시간(분),가게평점
3,STR004,불야성 강남점,야식,관악구,29,3.5
23,STR024,불야성 강남점,야식,영등포구,45,3.3
26,STR027,포장마차 강남점,야식,관악구,20,4.6
30,STR031,야식공작소 신촌점,야식,종로구,40,3.5
79,STR080,호프집 신촌점,야식,송파구,36,4.6
80,STR081,호프집 역삼점,야식,강남구,23,3.6
84,STR085,야식의민족 역삼점,야식,성동구,40,4.3
93,STR094,자담치킨 신촌점,야식,종로구,47,4.8
94,STR095,호프집 강남점,야식,영등포구,42,3.5


In [5]:
df_stores['지역'].unique()

array(['송파구', '관악구', '관악구  ', '서초구', '성동구', '관악', '영등포구', '종로구', '송파구  ',
       '마포구', '영등포', '강남구', '강남구  ', '종로구  ', '강남', '종로'], dtype=object)

In [6]:
df_customers

,고객ID,성별,연령,지역,가입일,등급
0,CUST0001,여,36.0,영등포구,2022-11-14,VIP
1,CUST0002,남,22.0,마포구,2023-03-02,일반
2,CUST0003,남,59.0,종로구,2023/12/10,일반
3,CUST0004,남,49.0,종로구,2022-01-18,일반
4,CUST0005,여,29.0,강남구,2022-09-03,실버
...,...,...,...,...,...,...
495,CUST0496,남,56.0,강남구,NaN,실버
496,CUST0497,여,39.0,서초구,2022-04-20,일반
497,CUST0498,여,38.0,서초구,2023-04-02,VIP
498,CUST0499,남,32.0,송파구,2024-06-14,일반


In [7]:
df_orders

,주문ID,주문일시,고객ID,가게ID,메뉴,주문금액,수량,배달소요시간(분),결제방법,쿠폰사용여부,평점,주문상태
0,ORD00001,2024-10-29 12:34:52,CUST0171,STR029,깐풍기,7300,1,56.0,카드,N,4.0,배달완료
1,ORD00002,2024-07-04 19:04:00,CUST0024,STR094,오돌뼈,25400,1,31.0,간편결제,N,NaN,배달완료
2,ORD00003,2024-08-03 19:00:58,CUST0425,STR024,닭볶음탕,33200,1,43.0,카드,N,NaN,배달중
3,ORD00004,2024-04-01 17:15:20,CUST0208,STR037,된장찌개,9000,2,35.0,카드,N,5.0,배달완료
4,ORD00005,2024-08-01 20:49:34,CUST0470,STR022,라멘,16700,1,60.0,카드,Y,NaN,배달완료
...,...,...,...,...,...,...,...,...,...,...,...,...
4995,ORD02157,2024-04-03 23:28:58,CUST0494,STR055,카페라떼,6800,1,42.0,카드,Y,NaN,배달완료
4996,ORD01565,2024-01-10 23:01:36,CUST0473,STR010,마늘치킨,18100,1,30.0,현금,Y,5.0,배달완료
4997,ORD01328,2024-02-28 13:35:23,CUST0408,STR084,시저샐러드,9100,1,31.0,카드,Y,3.0,배달완료
4998,ORD00919,2024-10-02 20:18:09,CUST0219,STR017,우동,19600,0,53.0,간편결제,N,4.0,배달완료


In [8]:
df_weather

,날짜,기온,강수량(mm),날씨
0,2024-01-01,0.6,0.0,흐림
1,2024-01-02,-2.2,0.0,흐림
2,2024-01-03,-2.6,0.0,흐림
3,2024-01-04,-3.3,0.0,맑음
4,2024-01-05,-0.6,0.0,맑음
...,...,...,...,...
361,2024-12-27,-2.0,0.0,흐림
362,2024-12-28,-1.9,0.0,흐림
363,2024-12-29,-2.4,0.0,흐림
364,2024-12-30,-1.6,0.0,맑음


In [9]:
df_weather['날씨'].unique()

array(['흐림', '맑음', '눈', '비'], dtype=object)

### TODO: 가게 데이터 미리보기

1. 구분선과 제목을 출력: `print("=" * 60)` → `print("[stores.csv] 가게 데이터")`
2. `display(stores.head())`로 상위 5행 확인
3. `stores.info()`로 컬럼 타입과 결측치 확인

**확인 포인트**: 가게ID, 가게명, 업종, 지역, 평균배달시간, 가게평점 컬럼이 있는지 확인


In [50]:
# 아래에 코드를 작성하세요
display( df_stores.head() )

,가게ID,가게명,업종,지역,평균배달시간(분),가게평점
0,STR001,맥도날드 본점,패스트푸드,송파구,24,3.1
1,STR002,프레시코드 홍대점,샐러드,송파구,30,4.4
2,STR003,하남돼지집 선릉점,일식,관악구,43,4.5
3,STR004,불야성 강남점,야식,관악구,29,3.5
4,STR005,신전떡볶이 선릉점,분식,송파구,42,4.6


In [51]:
df_stores.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   가게ID       100 non-null    object 
 1   가게명        100 non-null    object 
 2   업종         100 non-null    object 
 3   지역         100 non-null    object 
 4   평균배달시간(분)  100 non-null    int64  
 5   가게평점       95 non-null     float64
dtypes: float64(1), int64(1), object(4)
memory usage: 4.8+ KB


### TODO: 고객 데이터 미리보기

1. 구분선과 제목을 출력: `print("=" * 60)` → `print("[customers.csv] 고객 데이터")`
2. `display(customers.head())`로 상위 5행 확인
3. `customers.info()`로 컬럼 타입과 결측치 확인

**확인 포인트**: 고객ID, 성별, 연령, 지역, 가입일, 등급 컬럼 확인


In [53]:
# 아래에 코드를 작성하세요
df_customers.head()

,고객ID,성별,연령,지역,가입일,등급
0,CUST0001,여,36.0,영등포구,2022-11-14,VIP
1,CUST0002,남,22.0,마포구,2023-03-02,일반
2,CUST0003,남,59.0,종로구,2023/12/10,일반
3,CUST0004,남,49.0,종로구,2022-01-18,일반
4,CUST0005,여,29.0,강남구,2022-09-03,실버


In [54]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   고객ID    500 non-null    object 
 1   성별      477 non-null    object 
 2   연령      495 non-null    float64
 3   지역      500 non-null    object 
 4   가입일     473 non-null    object 
 5   등급      500 non-null    object 
dtypes: float64(1), object(5)
memory usage: 23.6+ KB


### TODO: 주문 데이터 미리보기

1. 구분선과 제목을 출력: `print("=" * 60)` → `print("[orders.csv] 주문 데이터")`
2. `display(orders.head())`로 상위 5행 확인
3. `orders.info()`로 컬럼 타입과 결측치 확인

**확인 포인트**: 주문금액이 `object` 타입(문자열)인지, 결측치가 어디에 있는지 확인


In [55]:
df_orders.head()

,주문ID,주문일시,고객ID,가게ID,메뉴,주문금액,수량,배달소요시간(분),결제방법,쿠폰사용여부,평점,주문상태
0,ORD00001,2024-10-29 12:34:52,CUST0171,STR029,깐풍기,7300,1,56.0,카드,N,4.0,배달완료
1,ORD00002,2024-07-04 19:04:00,CUST0024,STR094,오돌뼈,25400,1,31.0,간편결제,N,NaN,배달완료
2,ORD00003,2024-08-03 19:00:58,CUST0425,STR024,닭볶음탕,33200,1,43.0,카드,N,NaN,배달중
3,ORD00004,2024-04-01 17:15:20,CUST0208,STR037,된장찌개,9000,2,35.0,카드,N,5.0,배달완료
4,ORD00005,2024-08-01 20:49:34,CUST0470,STR022,라멘,16700,1,60.0,카드,Y,NaN,배달완료


In [56]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   주문ID       5000 non-null   object 
 1   주문일시       5000 non-null   object 
 2   고객ID       4995 non-null   object 
 3   가게ID       5000 non-null   object 
 4   메뉴         5000 non-null   object 
 5   주문금액       4997 non-null   object 
 6   수량         5000 non-null   int64  
 7   배달소요시간(분)  4996 non-null   float64
 8   결제방법       5000 non-null   object 
 9   쿠폰사용여부     5000 non-null   object 
 10  평점         2977 non-null   float64
 11  주문상태       5000 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 468.9+ KB


In [12]:
# 아래에 코드를 작성하세요


### TODO: 날씨 데이터 미리보기

1. 구분선과 제목을 출력: `print("=" * 60)` → `print("[weather.csv] 날씨 데이터")`
2. `display(weather.head())`로 상위 5행 확인
3. `weather.info()`로 컬럼 타입과 결측치 확인

**확인 포인트**: 날짜, 기온, 강수량, 날씨 컬럼 확인


In [57]:
df_weather.head()

,날짜,기온,강수량(mm),날씨
0,2024-01-01,0.6,0.0,흐림
1,2024-01-02,-2.2,0.0,흐림
2,2024-01-03,-2.6,0.0,흐림
3,2024-01-04,-3.3,0.0,맑음
4,2024-01-05,-0.6,0.0,맑음


In [58]:
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   날짜       366 non-null    object 
 1   기온       366 non-null    float64
 2   강수량(mm)  366 non-null    float64
 3   날씨       366 non-null    object 
dtypes: float64(2), object(2)
memory usage: 11.6+ KB


In [13]:
# 아래에 코드를 작성하세요


---
## 2. 데이터 품질 확인

분석 결과의 신뢰성을 확보하기 위해, 각 파일별로 **결측치, 중복, 이상값, 데이터 타입 문제**를 점검합니다.


### TODO: 결측치 현황 파악

4개 데이터프레임의 결측치를 한번에 점검하세요.

1. `for` 루프로 `('stores', stores), ('customers', customers), ('orders', orders), ('weather', weather)` 순회
2. 각 데이터프레임에서 `df.isnull().sum()`으로 결측치 합계 계산
3. 결측치가 0보다 큰 컬럼만 필터링: `missing = missing[missing > 0]`
4. 결측치가 있으면 컬럼명, 건수, 비율(%)을 출력
5. 없으면 "결측치 없음" 출력

사용할 메서드: `isnull()`, `sum()`

**기대 출력**: 각 데이터프레임별 결측치 컬럼과 건수/비율


In [61]:
missing = df_stores.isna().sum()

In [64]:
missing[ missing > 0 ]

가게평점    5
dtype: int64

In [67]:
df_stores.shape[0]

100

In [68]:
missing[ missing > 0 ] / df_stores.shape[0] * 100

가게평점    5.0
dtype: float64

In [71]:
missing = df_customers.isnull().sum()

In [74]:
missing[ missing > 0 ]

성별     23
연령      5
가입일    27
dtype: int64

In [76]:
missing[ missing > 0 ] / df_customers.shape[0] * 100

성별     4.6
연령     1.0
가입일    5.4
dtype: float64

In [78]:
missing =  df_orders.isna().sum()

In [81]:
missing[ missing > 0 ]

고객ID            5
주문금액            3
배달소요시간(분)       4
평점           2023
dtype: int64

In [82]:
missing[ missing > 0 ] / df_orders.shape[0] * 100

고객ID          0.10
주문금액          0.06
배달소요시간(분)     0.08
평점           40.46
dtype: float64

In [83]:
df_weather.isna().sum()

날짜         0
기온         0
강수량(mm)    0
날씨         0
dtype: int64

### TODO: 중복 데이터 확인

각 데이터프레임의 중복을 두 가지 관점에서 점검하세요.

1. **완전 중복**: `df.duplicated().sum()` — 모든 컬럼이 동일한 행
2. **ID 중복**: `df[id_col].duplicated().sum()` — ID 기준 중복

사용할 메서드: `duplicated()`, `sum()`

**기대 출력**: 각 데이터의 완전 중복 행 수, ID 중복 수


In [88]:
df_stores.duplicated().sum()

np.int64(0)

In [90]:
df_stores['가게ID'].duplicated().sum()

np.int64(0)

In [91]:
df_customers.duplicated().sum()

np.int64(0)

In [93]:
df_customers['고객ID'].duplicated().sum()

np.int64(0)

In [94]:
df_orders.duplicated().sum()

np.int64(5)

In [98]:
df_orders[ df_orders.duplicated() ]

,주문ID,주문일시,고객ID,가게ID,메뉴,주문금액,수량,배달소요시간(분),결제방법,쿠폰사용여부,평점,주문상태
4995,ORD02157,2024-04-03 23:28:58,CUST0494,STR055,카페라떼,6800,1,42.0,카드,Y,NaN,배달완료
4996,ORD01565,2024-01-10 23:01:36,CUST0473,STR010,마늘치킨,18100,1,30.0,현금,Y,5.0,배달완료
4997,ORD01328,2024-02-28 13:35:23,CUST0408,STR084,시저샐러드,9100,1,31.0,카드,Y,3.0,배달완료
4998,ORD00919,2024-10-02 20:18:09,CUST0219,STR017,우동,19600,0,53.0,간편결제,N,4.0,배달완료
4999,ORD00724,2024-05-26 22:09:37,CUST0121,STR038,오꼬노미야끼,10400,1,50.0,간편결제,Y,NaN,배달완료


In [100]:
df_orders['주문ID'].duplicated().sum()

np.int64(5)

In [102]:
df_orders[ df_orders['주문ID'] == 'ORD02157' ]

,주문ID,주문일시,고객ID,가게ID,메뉴,주문금액,수량,배달소요시간(분),결제방법,쿠폰사용여부,평점,주문상태
2156,ORD02157,2024-04-03 23:28:58,CUST0494,STR055,카페라떼,6800,1,42.0,카드,Y,NaN,배달완료
4995,ORD02157,2024-04-03 23:28:58,CUST0494,STR055,카페라떼,6800,1,42.0,카드,Y,NaN,배달완료


In [103]:
df_weather.duplicated().sum()

np.int64(0)

In [105]:
df_weather['날짜'].duplicated().sum()

np.int64(0)

In [15]:
# 아래에 코드를 작성하세요


### TODO: stores 이상값 확인

stores 데이터의 카테고리 값 분포를 확인하세요.

1. `stores['지역'].value_counts()` → 지역명이 통일되어 있는지 확인 (서울/서울시/서울특별시 등)
2. `stores['업종'].value_counts()` → 업종 고유값 확인

**확인 포인트**: 같은 지역이 다른 이름으로 표기되어 있지 않은지 체크


In [108]:
df_stores['지역'].unique()

array(['송파구', '관악구', '관악구  ', '서초구', '성동구', '관악', '영등포구', '종로구', '송파구  ',
       '마포구', '영등포', '강남구', '강남구  ', '종로구  ', '강남', '종로'], dtype=object)

In [107]:
df_stores['지역'].value_counts()

지역
송파구      20
서초구      12
관악구      10
성동구      10
영등포구     10
마포구      10
강남구      10
종로구       9
종로구       2
관악구       1
관악        1
송파구       1
영등포       1
강남구       1
강남        1
종로        1
Name: count, dtype: int64

In [110]:
df_stores['업종'].value_counts()

업종
일식        13
중식        12
샐러드       11
패스트푸드     11
피자        11
분식        11
치킨         9
야식         9
한식         8
카페/디저트     5
Name: count, dtype: int64

### TODO: customers 이상값 확인

customers 데이터의 값 분포를 확인하세요.

1. `customers['성별'].value_counts()` → 성별 표기가 통일되어 있는지 확인 (M/F, 남/여, 남성/여성 등)
2. `customers['연령'].min()`, `.max()`, `.describe()` → 연령 범위와 이상치 확인
3. `customers['등급'].value_counts()` → 등급 고유값 확인

**확인 포인트**: 비현실적 연령(0세 이하, 100세 이상), 성별 표기 불일치


In [111]:
df_customers['성별'].value_counts()

성별
남     227
여     221
M      17
남자     12
Name: count, dtype: int64

In [113]:
df_customers.describe()

,연령
count,495.000000
mean,38.686869
std,13.682820
min,0.000000
25%,28.000000
50%,38.000000
75%,50.000000
max,150.000000


In [115]:
df_customers['등급'].value_counts()

등급
일반     258
실버     120
골드      71
VIP     51
Name: count, dtype: int64

### TODO: orders 이상값 확인

orders 데이터의 값 분포와 타입 문제를 확인하세요.

1. **주문금액 dtype 확인**: `orders['주문금액'].dtype`
2. **문자열인 경우**: 쉼표(,)와 '원'을 제거한 후 숫자 변환을 시도하여 변환 불가 행 탐지
   1. `pd.to_numeric(errors='coerce')`로 숫자 변환
2. `replace()`로 값 치환
3. **주문상태 고유값**: `orders['주문상태'].value_counts()`
4. **결제방법 고유값**: `orders['결제방법'].value_counts()`

**기대 출력**: 주문금액의 dtype, 변환 불가 행 수, 주문상태/결제방법 분포


In [116]:
df_orders['주문금액']

0        7300
1       25400
2       33200
3        9000
4       16700
        ...  
4995     6800
4996    18100
4997     9100
4998    19600
4999    10400
Name: 주문금액, Length: 5000, dtype: object

In [122]:
pd.to_numeric(  df_orders['주문금액'].str.replace(',' , '').str.replace('원', '')  , errors='coerce' )

0        7300.0
1       25400.0
2       33200.0
3        9000.0
4       16700.0
         ...   
4995     6800.0
4996    18100.0
4997     9100.0
4998    19600.0
4999    10400.0
Name: 주문금액, Length: 5000, dtype: float64

In [123]:
df_orders['주문상태'].value_counts()

주문상태
배달완료    4252
취소       496
배달중      252
Name: count, dtype: int64

In [125]:
df_orders['결제방법'].value_counts()

결제방법
카드      2509
간편결제    2006
현금       485
Name: count, dtype: int64

In [18]:
# 아래에 코드를 작성하세요


### 데이터 품질을 검사해서 다음처럼 문제를 요약 했습니다.

| 파일 | 문제 유형 | 상세 내용 | 처리 방법 |
|------|----------|----------|----------|
| stores | 지역명 불일치 | 같은 지역이 다르게 표기 (예: '서울', '서울시', '서울특별시') | 통일 |
| stores | 결측치 | 일부 컬럼에 결측값 존재 | 적절한 값으로 대체 |
| customers | 성별 불일치 | 'M'/'F', '남'/'여', '남성'/'여성' 혼재 | '남성'/'여성'으로 통일 |
| customers | 연령 이상치 | 비현실적 연령값 (0세 이하, 100세 이상 등) | 범위 제한 또는 대체 |
| customers | 가입일 타입 | 문자열 → datetime 변환 필요 | pd.to_datetime |
| orders | 주문금액 타입 | 쉼표, '원' 등 문자 포함 가능 | 정제 후 숫자 변환 |
| orders | 중복/결측 | 중복 주문, 결측 평점 등 | 중복 제거, 결측 처리 |
| orders | 주문일시 타입 | 문자열 → datetime 변환 필요 | pd.to_datetime |


---
## 3. 데이터 전처리

발견된 품질 문제를 체계적으로 해결합니다.


### 3-1. stores 전처리


### TODO: stores 전처리

1. **원본 보존**: `stores_raw = stores.copy()`

2. **지역명 통일**: 공백 제거 후 매핑 딕셔너리로 대표 지역명 변환
   1. `replace()`로 값 치환

3. **결측치 처리**: `평균배달시간`, `가게평점` 컬럼에 결측이 있으면
   - 1차: 같은 `업종`의 중앙값으로 대체 → `groupby('업종')[col].transform('median')`
   - 2차: 그래도 남은 결측은 전체 중앙값으로 대체
   1. `fillna()`로 결측치 처리

4. **결과 출력**: 지역 고유값 수와 목록, 남은 결측치 수

**기대 출력**: 지역명이 통일되고 결측치 0건


In [126]:
stores_raw = df_stores.copy()

In [127]:
df_stores = stores_raw.copy()

In [128]:
df_stores['지역'].unique()

array(['송파구', '관악구', '관악구  ', '서초구', '성동구', '관악', '영등포구', '종로구', '송파구  ',
       '마포구', '영등포', '강남구', '강남구  ', '종로구  ', '강남', '종로'], dtype=object)

In [131]:
df_stores['지역'] = df_stores['지역'].str.strip()

In [132]:
df_stores['지역'].unique()

array(['송파구', '관악구', '서초구', '성동구', '관악', '영등포구', '종로구', '마포구', '영등포',
       '강남구', '강남', '종로'], dtype=object)

In [135]:
df_stores['지역'] = df_stores['지역'].replace( {'관악':'관악구', '영등포':'영등포구', '강남':'강남구', '종로':'종로구'})  

In [136]:
df_stores['지역'].unique()

array(['송파구', '관악구', '서초구', '성동구', '영등포구', '종로구', '마포구', '강남구'],
      dtype=object)

In [137]:
df_stores.isna().sum()

가게ID         0
가게명          0
업종           0
지역           0
평균배달시간(분)    0
가게평점         5
dtype: int64

In [141]:
df_stores.groupby('업종')['가게평점'].transform( 'median' )

0     3.95
1     4.20
2     4.00
3     3.60
4     4.40
      ... 
95    3.60
96    4.40
97    4.20
98    4.00
99    4.20
Name: 가게평점, Length: 100, dtype: float64

In [144]:
df_stores['가게평점'] = df_stores['가게평점'].fillna( df_stores.groupby('업종')['가게평점'].transform( 'median' )  )

In [145]:
df_stores.isna().sum()

가게ID         0
가게명          0
업종           0
지역           0
평균배달시간(분)    0
가게평점         0
dtype: int64

In [19]:
# 아래에 코드를 작성하세요


### 3-2. customers 전처리


### TODO: customers 전처리

1. **원본 보존**: `customers_raw = customers.copy()`

2. **성별 통일**: 다양한 표기를 `'남성'`, `'여성'`으로 통일
   1. `replace()`로 값 치환

3. **연령 이상치 처리**: 10세 미만 또는 80세 초과 → 중앙값으로 대체
   사용할 메서드: `median()`, `sum()`

4. **연령 결측치**: `customers['연령'] = customers['연령'].fillna(age_median)`

5. **가입일 변환**: `customers['가입일'] = pd.to_datetime(customers['가입일'], errors='coerce')`

6. **등급 결측**: `customers['등급'] = customers['등급'].fillna('일반')`

7. **지역명 통일**: stores와 동일한 `region_map`으로 변환
   1. `replace()`로 값 치환

8. **결과 출력**: 성별 고유값, 연령 범위, 이상치 처리 건수, 결측치 수

**기대 출력**: 성별 ['남성', '여성'], 연령 범위 정상, 결측치 0건


In [146]:
customers_raw = df_customers.copy()

In [147]:
df_customers = customers_raw.copy()

In [148]:
df_customers['성별'].unique()

array(['여', '남', nan, 'M', '남자'], dtype=object)

In [150]:
df_customers['성별'] = df_customers['성별'].replace( {'여':'여성', '남':'남성','M':'남성', '남자':'남성'} )

In [151]:
df_customers['성별'].unique()

array(['여성', '남성', nan], dtype=object)

In [152]:
df_customers['연령'].median()

38.0

In [162]:
df_customers.loc[ (df_customers['연령'] < 10 ) | (df_customers['연령'] > 80) , '연령'  ] = df_customers['연령'].median()

In [163]:
df_customers['연령'].describe()

count    495.000000
mean      38.537374
std       12.612976
min       18.000000
25%       28.000000
50%       38.000000
75%       50.000000
max       60.000000
Name: 연령, dtype: float64

In [20]:
# 아래에 코드를 작성하세요
aa

### 3-3. orders 전처리


### TODO: orders 전처리

orders 전처리는 여러 단계로 진행됩니다. 순서를 지켜주세요.

1. **원본 보존**: `orders_raw = orders.copy()`

2. **중복 제거**: `주문ID` 기준 첫 번째만 유지
   1. `drop_duplicates()`로 중복 제거

3. **주문금액 문자→숫자**: dtype이 `object`이면 쉼표(`,`)와 `'원'` 제거 후 `pd.to_numeric(errors='coerce')`
   1. `astype()`으로 타입 변환
2. `replace()`로 값 치환
3. `pd.to_numeric(errors='coerce')`로 숫자 변환

4. **주문금액 이상치**: 0 이하 또는 500,000 초과 → 중앙값으로 대체, 결측도 중앙값으로
   1. `fillna()`로 결측치 처리

5. **날짜/시간 변환 및 추출**:
   1. `pd.to_datetime()`으로 날짜 변환

6. **시간대 분류 함수**: `주문시간`을 아침/점심/오후/저녁/야식으로 분류
   사용할 메서드: `apply()`

7. **배달소요시간 처리**: 결측 → 중앙값, 0 이하 또는 120 초과 → 중앙값
   1. `fillna()`로 결측치 처리

8. **평점**: 결측치는 그대로 유지 (취소 주문 등은 평점이 없을 수 있음)

9. **수량 결측**: `orders['수량'] = orders['수량'].fillna(1)`

10. **쿠폰사용여부 통일**: 다양한 표기를 매핑
    1. `astype()`으로 타입 변환
2. `replace()`로 값 치환

11. **결과 출력**: 최종 행 수, 주문금액 범위, 주문기간, 남은 결측치

**기대 출력**: 주문금액이 숫자, 시간 관련 컬럼(주문날짜/시간/요일/월/시간대) 추가됨


In [21]:
# 아래에 코드를 작성하세요


### 3-4. weather 전처리


### TODO: weather 전처리

1. **날짜 변환**: `weather['날짜'] = pd.to_datetime(weather['날짜'], errors='coerce')`
2. **날짜 date 컬럼 추가**: `weather['날짜_date'] = weather['날짜'].dt.date`
3. **강수량 결측 → 0**: `weather['강수량'] = weather['강수량'].fillna(0)` (비가 안 온 날)
4. **날씨 공백 제거**: `weather['날씨'] = weather['날씨'].str.strip()`
5. **결과 출력**: 날씨 유형(sorted unique), 기간(min ~ max), 기온 범위

**기대 출력 예시**:
```
[weather 전처리 완료]
  날씨 유형: ['맑음', '비', '소나기', '눈', '흐림', ...]
  기간: 2024-01-01 ~ 2024-12-31
  기온 범위: -X.X°C ~ XX.X°C
```


In [22]:
# 아래에 코드를 작성하세요


---
## 4. 데이터 통합 (Merge)

분석을 위해 주문 데이터를 중심으로 가게, 고객, 날씨 정보를 결합합니다.


### TODO: 데이터 통합 (Merge)

orders를 중심으로 3번의 merge를 수행하세요.

1. **orders + stores**: `가게ID` 기준 left merge
   merge 순서:
1. `stores` 와 `on="가게ID"`, `how="left"` 로 merge


2. **+ customers**: `고객ID` 기준 left merge
   merge 순서:
1. `customers` 와 `on="고객ID"`, `how="left"` 로 merge


3. **+ weather**: 주문날짜와 날씨 날짜를 매칭
   merge 순서:
1. `weather` 와 `on="주문날짜_dt"`, `how="left"` 로 merge


4. **불필요 컬럼 제거**: `주문날짜_dt`, `날짜_merge`, `날짜_date` 컬럼 drop
   사용할 메서드: `drop()`

5. 최종 shape과 `df.head(3)`을 출력

**기대 출력**: 최종 데이터가 원래 orders 행 수와 동일, 열 수가 크게 증가


In [23]:
# 아래에 코드를 작성하세요


### TODO: 통합 데이터 검증

merge가 제대로 되었는지 확인하세요.

1. **가게 정보 매칭률**: `df['업종'].notna().sum() / len(df) * 100`
2. **고객 정보 매칭률**: `df['성별'].notna().sum() / len(df) * 100`
3. **날씨 정보 매칭률**: `df['날씨'].notna().sum() / len(df) * 100`
4. **연령대 컬럼 생성**: 고객 분석에서 사용할 연령대 파생 변수
   1. `astype()`으로 타입 변환
5. **컬럼 목록 출력**: `print(f"컬럼 목록: {list(df.columns)}")`

**기대 출력**: 매칭률이 모두 높음(90% 이상), 연령대 컬럼 추가됨


In [24]:
# 아래에 코드를 작성하세요


---
## 5. 매출 분석

전체 매출 추이와 지역별/업종별 매출 구조를 파악합니다.


### TODO: 핵심 매출 지표 계산

1. **취소 주문 제외**: `df_completed = df[df['주문상태'] != '취소'].copy()`
2. **총 매출**: `df_completed['주문금액'].sum()` → 변수 `total_revenue`에 저장
3. **평균 주문금액**: `df_completed['주문금액'].mean()` → 변수 `avg_order_value`에 저장
4. **총 주문건수**: `len(df_completed)` → 변수 `total_orders`에 저장
5. 분석 대상 건수, 총 매출, 총 주문건수, 평균 주문금액을 출력

**기대 출력**:
```
분석 대상: X,XXX건 (취소 XXX건 제외)
========================================
  총 매출:     XXX,XXX,XXX원
  총 주문건수:       X,XXX건
  평균 주문금액:      XX,XXX원
========================================
```


In [25]:
# 아래에 코드를 작성하세요


### TODO: 월별 매출 추이 시각화

1. **월별 집계** → 변수 `monthly`:
   사용할 메서드: `agg()`, `groupby()`, `reset_index()`

2. **이중 축 그래프 생성** (`figsize=(14, 6)`):
   - **왼쪽 축(ax1)**: 매출 막대 그래프 (만원 단위, `monthly['매출'] / 10000`)
     - `ax1.bar()`, `color=COLORS[0]`, `alpha=0.7`, `width=0.6`, `label='매출(만원)'`
     - x축 라벨: `[f'{m}월' for m in monthly['주문월']]`
   - **오른쪽 축(ax2)**: 주문건수 선 그래프
     - `ax2 = ax1.twinx()`
     - `ax2.plot()`, `color=COLORS[1]`, `marker='o'`, `linewidth=2.5`, `label='주문건수'`

3. **범례 합치기**:
   사용할 메서드: `get_legend_handles_labels()`, `legend()`

4. **제목**: `plt.title('월별 매출 추이 및 주문건수', fontsize=16, fontweight='bold', pad=15)`

**기대 출력**: 막대(매출)와 선(주문건수)이 함께 표시된 이중 축 그래프


In [26]:
# 아래에 코드를 작성하세요


### TODO: 지역별 매출 비교 시각화

1. **지역 컬럼 확인**: merge 시 suffixes로 인해 `'지역'` 또는 `'지역_가게'`일 수 있음
   

2. **지역별 집계**: `groupby(region_col)['주문금액'].agg(['sum', 'count', 'mean']).reset_index()`
   - 컬럼명 변경: `['지역', '총매출', '주문건수', '평균주문금액']`
   - 총매출 오름차순 정렬

3. **가로 막대 그래프** (`figsize=(12, 6)`):
   - `ax.barh(region_revenue['지역'], region_revenue['총매출'] / 10000, color=COLORS[:])`
   - 각 막대 끝에 값 표시: `f'{val/10000:,.0f}만원'`
   - x축 라벨: '매출 (만원)'
   - 제목: '지역별 총 매출 비교'

**기대 출력**: 지역별 총 매출을 보여주는 가로 막대 그래프


In [27]:
# 아래에 코드를 작성하세요


### TODO: 업종별 매출 & 주문건수 시각화

1. **업종별 집계** → 총매출 내림차순 정렬 → 변수 `biz_stats`:
   사용할 메서드: `agg()`, `groupby()`, `reset_index()`, `sort_values()`

2. **1행 2열 서브플롯** (`figsize=(16, 6)`):
   - **왼쪽(ax1)**: 업종별 총매출 세로 막대 그래프 (만원 단위)
     - 막대 위에 값 표시: `f'{val/10000:,.0f}'`
     - x축 라벨 45도 회전: `ax1.tick_params(axis='x', rotation=45)`
   - **오른쪽(ax2)**: 업종별 주문건수 비중 파이 차트
     - `ax2.pie(biz_stats['주문건수'], labels=biz_stats['업종'], autopct='%1.1f%%', startangle=90)`

3. **테이블 출력**:
   

**기대 출력**: 막대 그래프 + 파이 차트 + 지표 테이블


In [28]:
# 아래에 코드를 작성하세요


---
## 6. 시간대 분석

주문이 집중되는 시간대와 요일을 파악하여, 운영 효율화 및 마케팅 타이밍에 활용합니다.


### TODO: 시간대별 주문 패턴 시각화

1. **시간별 집계** → 변수 `hourly`:
   사용할 메서드: `count()`, `groupby()`, `reset_index()`

2. **시간대별 색상 분류** (`figsize=(14, 6)`):
   

3. **막대 그래프**: `ax.bar(hourly['시간'], hourly['주문건수'], color=bar_colors, edgecolor='white')`
   - x축: `ax.set_xticks(range(24))`, 라벨 `[f'{h}시' for h in range(24)]`, `rotation=45`

4. **피크 시간 표시**: 최대 주문건수 시간에 화살표 주석
   사용할 메서드: `annotate()`, `idxmax()`, `max()`

5. **커스텀 범례**: `from matplotlib.patches import Patch`로 점심/저녁/야식/기타 범례 생성

6. **제목**: '시간대별 주문 패턴'

**기대 출력**: 24시간 막대 그래프, 점심/저녁 피크가 색상으로 구분됨


In [29]:
# 아래에 코드를 작성하세요


### TODO: 요일별 주문 패턴 시각화

1. **요일 순서 정의**:
   

2. **요일별 집계**:
   사용할 메서드: `agg()`, `groupby()`, `reindex()`, `reset_index()`

3. **이중 축 그래프** (`figsize=(12, 6)`):
   - **막대(ax1)**: 주문건수, 주중(월~금)=`COLORS[0]`, 주말(토/일)=`COLORS[3]`
     사용할 메서드: `bar()`
   - **선(ax2=ax1.twinx())**: 평균주문금액, `marker='s'`, `color=COLORS[2]`, `linewidth=2.5`

4. **커스텀 범례**: 주중/주말/평균주문금액 3개
   사용할 메서드: `Line2D()`, `legend()`

5. **제목**: '요일별 주문 패턴'

**기대 출력**: 주중/주말이 색상으로 구분된 막대 + 평균주문금액 선 그래프


In [30]:
# 아래에 코드를 작성하세요


### TODO: 시간대 x 업종 히트맵

1. **교차표 생성** → 변수 `time_biz`:
   사용할 메서드: `count()`, `groupby()`, `unstack()`

2. **시간대 순서 지정**: `['아침', '점심', '오후', '저녁', '야식']` 순서로 reindex
   사용할 메서드: `reindex()`

3. **히트맵 시각화** (`figsize=(14, 6)`):
   사용할 메서드: `heatmap()`

4. **제목**: '시간대 x 업종별 주문건수 히트맵'

**기대 출력**: 시간대(행) x 업종(열) 히트맵, 셀에 주문건수 표시


In [31]:
# 아래에 코드를 작성하세요


---
## 7. 메뉴 & 업종 분석

업종별 특성과 인기 메뉴를 분석하여 경쟁력 있는 카테고리를 파악합니다.


### TODO: 업종별 평균 주문금액 & 배달시간 비교

1. **업종별 집계** → 평균주문금액 내림차순 → 변수 `biz_profile`:
   사용할 메서드: `agg()`, `groupby()`, `reset_index()`, `sort_values()`

2. **1행 2열 서브플롯** (`figsize=(16, 6)`):
   - **왼쪽(ax1)**: 업종별 평균 주문금액 가로 막대
     - `ax1.barh(biz_profile['업종'], biz_profile['평균주문금액'], color=COLORS[:])`
     - 막대 끝에 `f'{val:,.0f}원'` 표시
     - 제목: '업종별 평균 주문금액'
   - **오른쪽(ax2)**: 업종별 평균 배달시간 가로 막대 (평균배달시간 내림차순 재정렬)
     - `biz_profile_sorted = biz_profile.sort_values('평균배달시간', ascending=False)`
     - 막대 끝에 `f'{val:.1f}분'` 표시
     - 제목: '업종별 평균 배달시간'

**기대 출력**: 업종별 평균 주문금액과 배달시간을 비교하는 2개 가로 막대 그래프


In [32]:
# 아래에 코드를 작성하세요


### TODO: 인기 메뉴 Top 20 시각화

1. **메뉴별 주문건수** → 변수 `top_menus`:
   사용할 메서드: `head()`, `reset_index()`, `value_counts()`

2. **가로 막대 그래프** (`figsize=(14, 8)`):
   사용할 메서드: `barh()`, `invert_yaxis()`, `set_yticklabels()`, `set_yticks()`
   - 막대 끝에 `f'{val}건'` 표시

3. **제목**: '인기 메뉴 Top 20'

**기대 출력**: 상위 20개 메뉴의 주문건수를 보여주는 가로 막대 그래프


In [33]:
# 아래에 코드를 작성하세요


### TODO: 업종별 평점 분포 비교 (박스플롯)

1. **평점 있는 주문만 필터**: `df_rated = df_completed.dropna(subset=['평점'])`

2. **업종 정렬**: 평점 중앙값 기준 내림차순
   사용할 메서드: `groupby()`, `median()`, `sort_values()`

3. **박스플롯** (`figsize=(14, 6)`):
   사용할 메서드: `boxplot()`, `set_alpha()`, `set_facecolor()`
   - 전체 평균선: `ax.axhline(y=df_rated['평점'].mean(), color='gray', linestyle='--', alpha=0.5, label=f'전체 평균 {df_rated["평점"].mean():.2f}')`
   - x축 라벨 45도 회전

4. **제목**: '업종별 평점 분포 비교'

**기대 출력**: 업종별 평점 분포를 보여주는 박스플롯, 빨간 다이아몬드가 평균


In [34]:
# 아래에 코드를 작성하세요


---
## 8. 배달 성과 분석

배달 소요시간과 고객 만족도의 관계, 취소율 등을 분석하여 서비스 품질을 점검합니다.


### TODO: 배달소요시간 분포 시각화

1. **히스토그램** (`figsize=(14, 6)`):
   사용할 메서드: `hist()`

2. **통계선 표시** (변수로 저장 — 나중에 사용):
   사용할 메서드: `axvline()`, `mean()`, `median()`

3. **제목**: '배달소요시간 분포', 범례 표시

4. **통계 출력**:
   - 평균, 중앙값, 표준편차
   - 30분 이내 배달 비율: `(df_completed['배달소요시간'] <= 30).mean()*100`
   - 60분 이상 배달 비율: `(df_completed['배달소요시간'] >= 60).mean()*100`

**기대 출력**: 히스토그램 + 평균/중앙값 수직선 + 통계 수치


In [35]:
# 아래에 코드를 작성하세요


### TODO: 배달시간 vs 평점 관계 시각화

1. **평점 있는 데이터만**: `df_scatter = df_completed.dropna(subset=['평점'])`

2. **산점도** (`figsize=(12, 7)`):
   사용할 메서드: `scatter()`

3. **구간별 평균 추세선**:
   사용할 메서드: `apply()`, `cut()`, `groupby()`, `mean()`, `plot()`, `reset_index()`

4. **제목**: '배달소요시간과 평점의 관계'

5. **상관계수 계산** (나중에 사용하므로 변수 `corr`에 저장):
   사용할 메서드: `corr()`

**기대 출력**: 산점도 + 구간 평균 추세선, 상관계수 출력


In [36]:
# 아래에 코드를 작성하세요


### TODO: 지역별 평균 배달시간 비교

1. **지역별 집계**: `df_completed.groupby(region_col)['배달소요시간'].agg(['mean', 'median', 'count']).reset_index()`
   - 컬럼명: `['지역', '평균', '중앙값', '건수']`
   - 평균 오름차순 정렬

2. **가로 막대 그래프** (`figsize=(12, 6)`):
   - `ax.barh(region_delivery['지역'], region_delivery['평균'], color=COLORS[:], alpha=0.8)`
   - 막대 끝에 `f'{val:.1f}분'` 표시
   - **전체 평균선**: `ax.axvline(x=df_completed['배달소요시간'].mean(), color='red', linestyle='--', alpha=0.7, label=f'전체 평균 {df_completed["배달소요시간"].mean():.1f}분')`
   - 범례 표시

3. **제목**: '지역별 평균 배달소요시간'

**기대 출력**: 지역별 평균 배달시간 가로 막대 + 전체 평균 수직선


In [37]:
# 아래에 코드를 작성하세요


### TODO: 취소율 분석 시각화

**주의**: 취소율은 `df` (전체 데이터)를 사용합니다 (`df_completed`가 아님!).

1. **업종별 취소율**:
   사용할 메서드: `apply()`, `groupby()`, `mean()`, `reset_index()`, `sort_values()`

2. **시간대별 취소율**: 같은 방식으로 `시간대`별 집계
   - 시간대 순서: `pd.Categorical(categories=['아침', '점심', '오후', '저녁', '야식'], ordered=True)`

3. **1행 2열 서브플롯** (`figsize=(16, 6)`):
   - **왼쪽(ax1)**: 업종별 취소율 가로 막대, 막대 끝에 `f'{rate:.1f}%'`
   - **오른쪽(ax2)**: 시간대별 취소율 세로 막대, 막대 위에 `f'{rate:.1f}%'`

4. **전체 취소율 출력** (나중에 사용하므로 변수 저장):
   사용할 메서드: `mean()`

**기대 출력**: 업종별/시간대별 취소율 그래프 + 전체 취소율 수치


In [38]:
# 아래에 코드를 작성하세요


---
## 9. 고객 분석

고객 세그먼트별 구매 패턴을 파악하여 타겟 마케팅 전략을 수립합니다.


### TODO: 연령대별/성별 주문 패턴 시각화

1. **연령대-성별 집계**:
   사용할 메서드: `agg()`, `groupby()`, `reset_index()`

2. **연령대 정렬** (나중에 사용하므로 변수 저장):
   1. `replace()`로 값 치환

3. **1행 2열 서브플롯** (`figsize=(16, 6)`):
   - **왼쪽(ax1)**: 연령대별 주문건수 (성별 구분 막대)
     1. `fillna()`로 결측치 처리
     - 제목: '연령대별 · 성별 주문건수', x축 회전 0도
   - **오른쪽(ax2)**: 연령대별 평균 주문금액 (성별 구분 막대)
     - 같은 pivot 방식, `values='평균주문금액'`
     - 제목: '연령대별 · 성별 평균 주문금액'

**기대 출력**: 연령대별 성별 주문건수/평균금액 비교 그래프


In [39]:
# 아래에 코드를 작성하세요


### TODO: 고객 등급별 구매 행동 비교

1. **등급별 집계** → 총매출 내림차순 → 변수 `grade_stats`:
   사용할 메서드: `agg()`, `groupby()`, `mean()`, `reset_index()`, `sort_values()`

2. **테이블 출력**:
   

3. **1행 2열 서브플롯** (`figsize=(14, 5)`):
   - **왼쪽(ax1)**: 등급별 매출 비중 파이 차트
     - `ax1.pie(grade_stats['총매출'], labels=grade_stats['등급'], autopct='%1.1f%%', startangle=90)`
   - **오른쪽(ax2)**: 등급별 평균 주문금액 세로 막대
     - 막대 위에 `f'{val:,.0f}원'` 표시

**기대 출력**: 등급별 지표 테이블 + 파이 차트 + 막대 그래프


In [40]:
# 아래에 코드를 작성하세요


### TODO: 재주문률 분석

1. **고객별 주문횟수**:
   사용할 메서드: `count()`, `groupby()`, `reset_index()`

2. **1행 2열 서브플롯** (`figsize=(16, 6)`):
   - **왼쪽(ax1)**: 주문횟수 히스토그램
     사용할 메서드: `hist()`, `max()`
     - 제목: '고객별 주문횟수 분포'
   - **오른쪽(ax2)**: 재주문 고객 비율 파이 차트
     사용할 메서드: `pie()`, `sum()`
     - 제목: '재주문 고객 비율'

3. **통계 출력** (변수 저장 — 나중에 사용):
   사용할 메서드: `max()`, `mean()`

**기대 출력**: 주문횟수 분포 히스토그램 + 재주문 비율 파이 차트 + 통계


In [41]:
# 아래에 코드를 작성하세요


### TODO: 쿠폰 사용 효과 분석

1. **쿠폰 사용 여부별 집계**:
   사용할 메서드: `agg()`, `groupby()`, `map()`, `reset_index()`

2. **1행 3열 서브플롯** (`figsize=(16, 5)`):
   - 3개 지표를 각각 세로 막대 그래프로:

   | 차트 | 지표 | 단위 | 제목 |
   |------|------|------|------|
   | 1 | 평균주문금액 | 원 | 쿠폰 사용 여부별 평균 주문금액 |
   | 2 | 평균평점 | 점 | 쿠폰 사용 여부별 평균 평점 |
   | 3 | 평균배달시간 | 분 | 쿠폰 사용 여부별 평균 배달시간 |

   - 막대 색상: `[COLORS[0], COLORS[2]]`
   - 막대 위에 값 표시 (원: `f'{val:,.0f}원'`, 점: `f'{val:.2f}점'`, 분: `f'{val:.1f}분'`)

3. **전체 쿠폰 사용률** (나중에 사용하므로 변수 저장):
   사용할 메서드: `mean()`

**기대 출력**: 쿠폰 사용/미사용 비교 3개 그래프 + 전체 쿠폰 사용률


In [42]:
# 아래에 코드를 작성하세요


---
## 10. 날씨 영향 분석

날씨가 배달 주문에 미치는 영향을 분석합니다. 기상 조건에 따른 수요 예측에 활용할 수 있습니다.


### TODO: 날씨별 주문건수 비교

1. **날씨 있는 데이터만**: `weather_orders = df_completed.dropna(subset=['날씨'])`

2. **날씨별 총 주문건수**:
   사용할 메서드: `count()`, `groupby()`, `reset_index()`, `sort_values()`

3. **날씨별 일평균 주문건수** (공정한 비교를 위해):
   사용할 메서드: `groupby()`, `merge()`, `reset_index()`, `size()`, `sort_values()`

4. **1행 2열 서브플롯** (`figsize=(16, 6)`):
   - **왼쪽(ax1)**: 총 주문건수 가로 막대
     - 값 표시: `f'{val:,}건'`
     - 제목: '날씨별 총 주문건수'
   - **오른쪽(ax2)**: 일평균 주문건수 가로 막대
     - 값 표시: `f'{val:.1f}건'`
     - 제목: '날씨별 일평균 주문건수'

**기대 출력**: 날씨별 총 주문건수 + 일평균 주문건수 비교 그래프


In [43]:
# 아래에 코드를 작성하세요


### TODO: 기온/강수량 vs 주문건수 시각화

1. **일별 주문 집계**:
   1. `pd.to_datetime()`으로 날짜 변환

2. **날씨 데이터와 결합**: `daily_weather = daily_orders.merge(weather, left_on='주문날짜', right_on='날짜', how='left')`

3. **1행 2열 서브플롯** (`figsize=(16, 6)`):
   - **왼쪽(ax1)**: 기온 vs 주문건수 산점도 + 1차 추세선
     1. `dropna()`로 결측 행 제거
     - 제목: '기온과 주문건수의 관계'
   - **오른쪽(ax2)**: 강수량 vs 주문건수 산점도 + 추세선 (강수량 > 0인 날만)
     - 제목: '강수량과 주문건수의 관계'

4. **상관계수 계산** (변수 저장 — 나중에 사용):
   사용할 메서드: `corr()`

**기대 출력**: 기온/강수량 산점도 + 추세선, 상관계수 출력


In [44]:
# 아래에 코드를 작성하세요


### TODO: 비 오는 날 업종별 주문 변화

1. **비/눈 여부 구분**:
   사용할 메서드: `isin()`, `map()`

2. **교차표**: `rain_biz = weather_orders.groupby(['비여부', '업종'])['주문ID'].count().unstack(fill_value=0)`

3. **일평균으로 비교** (각 날씨 상태의 일수로 나누기):
   사용할 메서드: `isin()`

4. **변화율 계산**:
   사용할 메서드: `DataFrame()`, `sort_values()`

5. **1행 2열 서브플롯** (`figsize=(16, 6)`):
   - **왼쪽(ax1)**: 업종별 일평균 주문 비교 (비/눈 vs 맑음/흐림 나란히)
     사용할 메서드: `arange()`, `barh()`, `set_yticklabels()`, `set_yticks()`
     - 제목: '날씨별 업종 일평균 주문'
   - **오른쪽(ax2)**: 업종별 변화율 가로 막대
     - 양수(빨강 `COLORS[3]`), 음수(파랑 `COLORS[0]`) 색상 구분
     - `ax2.axvline(x=0, color='black', linewidth=0.5)` 기준선
     - 값 표시: `f'{val:+.1f}%'`
     - 제목: '비/눈 올 때 주문 변화율'

**기대 출력**: 날씨별 업종 주문 비교 + 변화율 그래프


In [45]:
# 아래에 코드를 작성하세요


---
## 11. 종합 인사이트 & 제안


### TODO: 핵심 지표 요약 출력

지금까지 계산한 핵심 변수들을 한 화면에 요약 출력하세요.

출력할 변수들 (이전 셀에서 저장한 변수):
- **매출**: `total_revenue`, `total_orders`, `avg_order_value`
- **배달**: `mean_time`, 30분 내 배달률, `corr` (배달시간-평점 상관)
- **고객**: `total_customers`, `reorder_rate`, `coupon_rate`
- **기타**: `total_cancel_rate`, `corr_temp`, `corr_rain`

출력 형식:
사용할 메서드: `mean()`


In [46]:
# 아래에 코드를 작성하세요


### TODO: 최종 대시보드 (2x3 서브플롯)

핵심 결과 6개를 한 화면에 요약하는 대시보드를 만드세요.

**레이아웃**: `fig, axes = plt.subplots(2, 3, figsize=(20, 12))`

| 위치 | 차트 | 데이터 |
|------|------|--------|
| (0,0) | 월별 매출 추이 | `monthly['매출'] / 10000` 막대, `color=COLORS[0]` |
| (0,1) | 업종별 매출 Top | `biz_stats.head(6)` 가로 막대 |
| (0,2) | 시간대별 주문 패턴 | `hourly` 막대, `color=COLORS[2]` |
| (1,0) | 배달시간 분포 | 히스토그램 `bins=25` + 평균선(빨강 점선) |
| (1,1) | 연령대별 주문건수 | `groupby('연령대')['주문ID'].count().reindex(age_order)` 막대 |
| (1,2) | 날씨별 일평균 주문 | `weather_avg` 일평균주문 오름차순 가로 막대 |

각 차트에 제목, x/y 라벨을 설정하세요.

사용할 메서드: `savefig()`, `show()`, `suptitle()`, `tight_layout()`

**기대 출력**: 6개 차트가 배치된 종합 대시보드 + 이미지 저장


In [47]:
# 아래에 코드를 작성하세요


---
## 비즈니스 제안

데이터 분석 결과를 바탕으로 다음과 같은 비즈니스 제안을 도출합니다.

### 1. 매출 성장 전략
- **피크 시간대 활용**: 점심(11~14시)과 저녁(17~21시)에 주문이 집중되므로, 이 시간대에 맞춘 프로모션과 광고를 강화합니다.
- **주말 타겟 마케팅**: 주말에 주문량이 증가하는 패턴을 활용하여, 금요일 오후부터 주말 특가 이벤트를 운영합니다.
- **비수기 매출 보완**: 주문이 적은 시간대(아침, 오후)에 할인 쿠폰을 발행하여 수요를 분산시킵니다.

### 2. 배달 서비스 개선
- **배달시간 단축**: 배달 소요시간과 평점이 음의 상관관계를 보이므로, 배달시간 단축이 고객 만족도 향상에 직접적으로 기여합니다.
- **배달 지연 지역 집중 관리**: 평균 배달시간이 높은 지역에 라이더 배치를 늘리거나, 가게-고객 간 거리 최적화를 진행합니다.
- **취소율 낮추기**: 취소율이 높은 업종과 시간대를 분석하여, 예상 배달시간을 정확하게 안내하고 대기 시간 알림을 강화합니다.

### 3. 고객 관리 전략
- **VIP 고객 관리**: 상위 등급 고객의 평균 주문금액이 높으므로, 맞춤형 혜택(무료 배달, 전용 쿠폰)으로 이탈을 방지합니다.
- **재주문률 향상**: 1회 주문 고객에게 재주문 유도 쿠폰을 발송하여, 충성 고객으로 전환합니다.
- **쿠폰 전략 최적화**: 쿠폰 사용 시 주문금액 변화를 모니터링하여, ROI가 높은 쿠폰 타입을 발굴합니다.

### 4. 날씨 대응 전략
- **우천 시 프로모션**: 비/눈 오는 날 주문이 늘어나는 업종에 대해 맞춤 프로모션을 자동 발동합니다.
- **기상 기반 수요 예측**: 기온, 강수량 데이터를 활용하여 일별 주문량을 예측하고, 라이더 수급을 사전에 조절합니다.
- **계절별 메뉴 추천**: 기온에 따라 인기 메뉴가 달라지므로, 계절별 추천 메뉴 큐레이션을 도입합니다.

---

**분석 완료** | 본 보고서는 가상 데이터를 기반으로 작성되었습니다.
